htdocs/core/lib/files.lib.php

```diff
@@ -229,12 +229,17 @@ function dol_dir_list($path, $types="all", $recursive=0, $filter="", $excludefil
 function dol_dir_list_in_database($path, $filter="", $excludefilter=null, $sortcriteria="name", $sortorder=SORT_ASC, $mode=0)
 {
 	global $conf, $db;
+	global $object;
 
 	$sql=" SELECT rowid, label, entity, filename, filepath, fullpath_orig, keywords, cover, gen_or_uploaded, extraparams, date_c, date_m, fk_user_c, fk_user_m, acl, position";
 	if ($mode) $sql.=", description";
 	$sql.=" FROM ".MAIN_DB_PREFIX."ecm_files";
 	$sql.=" WHERE filepath = '".$db->escape($path)."'";
-	$sql.=" AND entity = ".$conf->entity;
+	if (!empty($object->element)) {
+		$sql.=" AND entity IN (".getEntity($object->element).")";
```

Comment:
If we do that, the "dol_dir_list_in_database" may return record that are in another entity.
The entity is inside the unique key of llx_ecm_file so it is possible to have each entity to have it own "index" of files.
So i see no reason to list here entries that are in another entity.
When do you get the duplicate key error ?

Reply:
@eldy 
I have this error when for example I have a product in the main entity, I upload a document for this product, it creates the record in this table, but if I share the product database with another entity and as I go on the document tab of this product, it tries to create another record in this table because it does not find it with this entity ID. There should not be a duplicate!

Reply:
@eldy 
you tested?

Reply:
You say
"...I upload a document for this product, it creates the record in this table, but if I share the product database with another entity and as I go on the document tab of this product, it tries to create another record in this table because it does not find it with this entity ID. There should not be a duplicate!"
And yes, this is exactly what we want :
"...I upload a document for this product, it creates the record in this table, but if I share the product database with another entity and as I go on the document tab of this product, it tries to create another record in this table because it does not find it with this entity ID."
Currently this is what we want.
If we have the field "entity" into table llx_ecm_files, it means we want to have the indexation into entity 2 that can differs from indexation into entity 1 so we should get 2 entries (one for each entity). The question is why do you get a duplicate error. You should not because the unique key include the entity. Do you have the "entity" field into the constraint unique key ?

Reply:
@eldy 
tu as une autre solution ?

Reply:
Migration 5 to 6 did this
ALTER TABLE llx_ecm_files DROP INDEX uk_ecm_files;
ALTER TABLE llx_ecm_files ADD UNIQUE INDEX uk_ecm_files (filepath, filename, entity);

But for a database created from scratch we don't have it.

It is the expected behavior fo have unknown in gen_or_upload and fullpath_orig.
The file is upload in entity 1, so dolibarr know this file is uploaded and from which file name.

If we switch on entity 2, system see a file stored on disk and readable, so file is indexed to for entity 2 automatically. But it is another entity, so we must have file indexed but information of "the file name on computer of the guy that upload the file" must not be visible. It should remain "unknown" for entity 2.
If you remove the file from entity 2, the file is completely "deleted" from disk, so it must disappear from 2. So there is a bug here. It does not delete the index in 1 because the product was in 1 but because the delete delete randomly the first one. And this is another bug (no feature must works randomly in Dolibarr).
I pushed a fix for both this problem in 8.0